### PR TITLE
Distr fix

### PR DESCRIPTION
**Pull request by hliu26**

### CODE REVIEW

File: symbulate/distributions.py [hunk 1]
```diff
@@ -618,16 +657,24 @@ def __init__(self,
...
+        if (sd2 < 0):
+            raise Exception("sd2 cannot be less than 0")
+
+
         if var1 is None:
             var1 = sd1 ** 2
         if var2 is None:
             var2 = sd2 ** 2
+        if (var1 < 0) or (var2 < 0):
+            raise Exception("var1 or var2 cannot be negative")
```
Comment (dlsun) [inline symbulate/distributions.py]: "var1 and var2 cannot be negative"

File: symbulate/distributions.py [hunk 2]
```diff
@@ -618,16 +657,24 @@ def __init__(self,
...
         if cov is None:
             if var1 * var2 < 0:
```
Comment (dlsun) [inline symbulate/distributions.py]: I don't get it. This check seems unnecessary to me. You've already verified above that var1 and var2 are both greater than 0, so of course var1 * var2 are also > 0.

File: symbulate/distributions.py [hunk 3]
```diff
@@ -94,10 +94,13 @@ class Binomial(Distribution):
...
         
-        if n >= 0 and (isinstance(n, int)):
+        if n > 0 and (isinstance(n, int)):
             self.n = n
+        elif n == 0:
+            raise NotImplementedError
+            #TODO
         else:
-            raise Exception("n must be an integer greater than or equal to 0")
+            raise Exception("n must be an integer greater than 0")
```
Comment (dlsun) [inline symbulate/distributions.py]: "greater than or equal to 0"

File: symbulate/distributions.py [hunk 4]
```diff
@@ -135,17 +138,17 @@ def __init__(self, n, N0, N1):
...
         else:
-            raise Exception("Number of draws must be an integer and cannot be negative")
+            raise Exception("n must be a positive integer")
         
-        if N0 > 0 and (isinstance(N0, int)):
+        if N0 >= 0 and (isinstance(N0, int)):
             self.N0 = N0
         else:
-            raise Exception("Number of failures must be an integer and cannot be negative")
+            raise Exception("Number of 0s must be a non-negative integer")
```
Comment (dlsun) [inline symbulate/distributions.py]: I would call this `N0`.

File: symbulate/distributions.py [hunk 5]
```diff
@@ -135,17 +138,17 @@ def __init__(self, n, N0, N1):
...
         
-        if N1 > 0 and (isinstance(N1, int)):
+        if N1 >= 0 and (isinstance(N1, int)):
             self.N1 = N1
         else:
-            raise Exception("Number of successes must be an integer and cannot be negative")
+            raise Exception("Number of 1s must be a non-negative integer")
```
Comment (dlsun) [inline symbulate/distributions.py]: Ditto, call this `N1`.

File: symbulate/distributions.py [hunk 6]
```diff
@@ -318,8 +321,8 @@ def __init__(self, a=0.0, b=1.0):
             "scale" : b - a
             }
 
-        if (b-a) <= 0:
-            raise Exception("b-a cannot be less than or equal to 0")
+        if (a > b):
```
Comment (dlsun) [inline symbulate/distributions.py]: No parentheses.

File: symbulate/distributions.py [hunk 7]
```diff
@@ -418,13 +443,19 @@ def __init__(self, shape, rate=1.0, scale=None):
...
             raise Exception("shape parameter cannot be less than or equal to 0")
         
-        if 0 < scale:
-            self.scale = scale
+        if scale is None:
+            if rate > 0:
+                self.rate = rate
+                self.scale = scale
+            else:
+                raise Exception("rate cannot be less than or equal to 0")
```
Comment (dlsun) [inline symbulate/distributions.py]: "...must be positive" or "must be greater than 0"

File: symbulate/distributions.py [hunk 8]
```diff
@@ -418,13 +443,19 @@ def __init__(self, shape, rate=1.0, scale=None):
...
         else:
-            raise Exception("scale parameter cannot be less than or equal to 0")
+            if scale > 0:
+                self.scale = scale
+                self.rate = 1. / scale
+            else:
+                raise Exception("scale cannot be less than or equal to 0")
```
Comment (dlsun) [inline symbulate/distributions.py]: Ditto as above.

File: symbulate/distributions.py [hunk 9]
```diff
@@ -341,14 +344,30 @@ class Normal(Distribution):
     """
 
     def __init__(self, mean=0.0, var=1.0, sd=None):
+
+        #Note: cleaner way to implement this
+
         if sd is None:
-            self.scale = np.sqrt(var)
+            if (var > 0):
+                self.var = var
```
Comment (dlsun) [inline symbulate/distributions.py]: No need to set self.var. It gets overwritten later when the parent __init__ method is called.

File: symbulate/distributions.py [hunk 10]
```diff
@@ -341,14 +344,30 @@ class Normal(Distribution):
     """
```
Comment (dlsun) [inline symbulate/distributions.py]: FYI (no change needed): this docstring is actually wrong at this point and will need to be modified sometime down the road.

File: symbulate/distributions.py [hunk 11]
```diff
@@ -536,12 +567,12 @@ def __init__(self, dfN, dfD):
         if dfN > 0:
             self.dfN = dfN
         else:
-            raise Exception("Degrees of freedom in numerator cannot be less than or equal to 0")
+            raise Exception("Degrees of freedom in numerator must be greater than 0")
```
Comment (dlsun) [inline symbulate/distributions.py]: I would write this as "dfN (numerator degrees of freedom) must be greater than 0".

Ditto for dfD below.

**Comment by dlsun** [comment]
Once you address my one last comment, please re-run all of your Jupyter test notebooks (for all of the distributions) from start to finish -- and verify that they produce the expected output. Then, let me know, and I will merge this pull request.

File: symbulate/distributions.py [hunk 12]
```diff
@@ -342,14 +342,14 @@ class Normal(Distribution):
...
     """
+    #TODO edit docstring for Normal Distribution
 
     def __init__(self, mean=0.0, var=1.0, sd=None):
 
         #Note: cleaner way to implement this
 
         if sd is None:
             if (var > 0):
-                self.var = var
```
Comment (dlsun) [inline symbulate/distributions.py]: You didn't delete self.var below. In general, when I make a comment, you need to fix the issue everywhere in the code, not just the line where I made the comment.

File: symbulate/distributions.py [hunk 13]
```diff
@@ -342,14 +342,14 @@ class Normal(Distribution):
...
         distribution (if specified, var parameter will be ignored)
     """
+    #TODO edit docstring for Normal Distribution
 
     def __init__(self, mean=0.0, var=1.0, sd=None):
 
         #Note: cleaner way to implement this
 
         if sd is None:
             if (var > 0):
```
Comment (dlsun) [inline symbulate/distributions.py]: No parentheses.

**Comment by hliu26** [comment]
Hi Prof,

I’ve re-run all the notebooks from start to finish and it would appear that they have all produced the expected output. I also got rid of extra parentheses, fixed some wordiness to error messages as well.

Thanks,

Howard

File: symbulate/distributions.py [hunk 14]
```diff
@@ -618,21 +653,26 @@ def __init__(self,
...
+        if sd2 < 0:
+            raise Exception("sd2 cannot be less than 0")
+
+        
         if var1 is None:
             var1 = sd1 ** 2
         if var2 is None:
             var2 = sd2 ** 2
+        if var1 < 0 or var2 < 0:
+            raise Exception("var1 and var 2 cannot be negative")
```
Comment (dlsun) [inline symbulate/distributions.py]: Fix typo.

File: symbulate/distributions.py [hunk 15]
```diff
@@ -570,19 +597,27 @@ class MultivariateNormal(Distribution):
...
+            self.mean = mean 
+        else:
+            raise Exception("Mean vector and Cov matrix cannot be empty")      
+ 
         if len(cov) >= 1:
-            if (all(len(row) == len(mean) for row in cov)):
-                self.cov = cov
+            if all(len(row) == len(mean) for row in cov):
+                try:
+                    if np.all(np.linalg.eigvals(matrix) >= 0):
```
Comment (dlsun) [inline symbulate/distributions.py]: Fix bug.

File: symbulate/distributions.py (hunk 14 above)
Comment (hliu26) [inline symbulate/distributions.py]: @dlsun I'm not sure what the typo is.

Comment (hliu26) [inline symbulate/distributions.py]: nevermind

**Pull request merged**